### PR TITLE
[BugFix] Peer reconnect timeout throwing NPE on destroyed.

### PR DESCRIPTION
**Pull request by rhyswilliamsza**
**What is the purpose of this pull request? (put an "X" next to item)**

[  ] Documentation update
[x] Bug fix
[ ] New feature
[ ] Other, please explain:

**What changes did you make? (Give an overview)**

Very small change to the handling of reconnections. When a torrent is destroyed, there is no current way of cancelling all outstanding peer reconnection attempts. This is due to the scoped design of the reconnection timeout reference. Thus, in the case where a torrent is destroyed, reconnection attempts still take place. When this happens, a NPE is thrown due to a missing `client` reference. Ideally, one would be able to cancel all outstanding reconnection attempts, however as described above, this is more challenging than anticipated. As an alternative solution, this simply checks whether the torrent is destroyed, and chokes the reconnection attempt at the start.

**Which issue (if any) does this pull request address?**
Described above. Referenced in https://github.com/webtorrent/webtorrent/issues/1932

**Is there anything you'd like reviewers to focus on?**
If there's a better way of cancelling the `reconnectTimeout` upon `_destroy()`, please let me know! :)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "webtorrent",
   "description": "Streaming torrent client",
-  "version": "0.108.6",
+  "version": "0.108.7",
```
Comment (hicom150) [inline package.json:4]: If you don't mind, please undo this change as it will be automatically cincremented with the release process we use 😉

File: lib/torrent.js [hunk 2]
```diff
@@ -1730,6 +1730,7 @@ class Torrent extends EventEmitter {
       )
 
       const reconnectTimeout = setTimeout(() => {
+        if (this.destroyed) return
         const newPeer = this._addPeer(peer.addr, this.client.utp ? 'utp' : 'tcp')
         if (newPeer) newPeer.retries = peer.retries + 1
       }, ms)
```
Comment (hicom150) [inline lib/torrent.js:1736]: Although this check will certainly fix the problem, as you have suggested I would change the `reconnectTimeout` reference scope to class level, in the same way that it is done with `_rechokeIntervalId`. So this code block could be something like this:
```
this._reconnectTimeoutId = setTimeout(() => {
  const newPeer = this._addPeer(peer.addr, this.client.utp ? 'utp' : 'tcp')
  if (newPeer) newPeer.retries = peer.retries + 1
}, ms)
if (this._reconnectTimeoutId.unref) this._reconnectTimeoutId.unref()
```
and we will also need to add in line 89
```
this._reconnectTimeoutId = null
```
and also add in line 661
```
clearTimeout(this._reconnectTimeoutId)
```

**Review by hicom150 — CHANGES_REQUESTED** [review]
Hey @rhyswilliamsza thank you for this great PR, it solves the problem perfectly 👍

As you correctly suggested maybe there could be a more "elegant" way to deal with the cancelling of `reconnectTimeout` so I made a small suggestion 😇

Let me know what do you think 😉

**Comment by hicom150** [comment]
@rhyswilliamsza any chance we can push forward this fix? 😉 I think it is an important one 👍

**Comment by rhyswilliamsza** [comment]
Hi @hicom150! Yes of course, sorry. Have been crazy busy with work. Will have a go this evening!

File: lib/torrent.js (hunk 2 above)
Comment (rhyswilliamsza) [reply lib/torrent.js]: Haha, sorry for being busy and blocking this. You should definitely have just run with the changes! xD

**Comment by rhyswilliamsza** [comment]
Sorry for the holdup on the rather simple change. One last thing. Line 1718:

```
      if (peer.retries >= RECONNECT_WAIT.length) {
        if (this.client.utp) {
          const newPeer = this._addPeer(peer.addr, 'tcp')
          if (newPeer) newPeer.retries = 0 <------------------------- 1718
        } else {
          this._debug(
            'conn %s closed: will not re-add (max %s attempts)',
            peer.addr, RECONNECT_WAIT.length
          )
          return
        }
      }
```

Should we not have a return here? It seems like this will attempt to add a TCP peer in the case of UTP, and if successful, immediately jump out and attempt a reconnection after 1s. This reconnection would likely die in `addPeer()` as a duplicate, however it seems a little odd?

**Comment by hicom150** [comment]
> Should we not have a return after this?

You are absolutely right! The `return` of the line `1722` should be placed outside the `if else` so it "returns" in both cases 😅

```
if (peer.retries >= RECONNECT_WAIT.length) {
  if (this.client.utp) {
    const newPeer = this._addPeer(peer.addr, 'tcp')
    if (newPeer) newPeer.retries = 0
  } else {
    this._debug(
      'conn %s closed: will not re-add (max %s attempts)',
      peer.addr, RECONNECT_WAIT.length
    )
  }

  return
}
```

File: package.json [hunk 3]
```diff
@@ -154,4 +154,4 @@
       "webtorrent.chromeapp.js"
     ]
   }
-}
```
Comment (hicom150) [inline package.json:157]: I think that we should leave this newline at the end 😉

**Review by hicom150 — APPROVED** [review]

File: package.json (hunk 3 above)
Comment (rhyswilliamsza) [reply package.json]: Bizarre! No clue why that happened. Thanks!

**Comment by rhyswilliamsza** [comment]
> > Should we not have a return after this?
> 
> You are absolutely right! The `return` of the line `1722` should be placed outside the `if else` so it "returns" in both cases 😅
> 
> ```
> if (peer.retries >= RECONNECT_WAIT.length) {
>   if (this.client.utp) {
>     const newPeer = this._addPeer(peer.addr, 'tcp')
>     if (newPeer) newPeer.retries = 0
>   } else {
>     this._debug(
>       'conn %s closed: will not re-add (max %s attempts)',
>       peer.addr, RECONNECT_WAIT.length
>     )
>   }
> 
>   return
> }
> ```

Awesome, thought as much. Have made the change :)

**Comment by hicom150** [comment]
Nice work @rhyswilliamsza 💪 It's been a pleasure helping you with this PR 😉

**Comment by feross** [comment]
Nice work @rhyswilliamsza! Thanks @hicom150 for shepherding this PR 👍 

Released as 0.109.1.

**Comment by feross** [comment]
So taking a closer look at this PR, it doesn't seem like it works if there is more than one peer that times out since `this._reconnectTimeoutId` will just get overwritten the next time we call `setTimeout`. I think the cleanest way to solve this was probably the original approach, even though it leaves a few timers around for a few seconds (not a big deal, since they're `unref`'d which means they won't keep the process alive.

**Comment by feross** [comment]
Sent a PR: https://github.com/webtorrent/webtorrent/pull/1944

**Comment by hicom150** [comment]
I'm sorry I pointed in the wrong direction 😓
I've approved and merged #1944 😉

**Comment by feross** [comment]
Thanks for reviewing the PR, @hicom150. Released as 0.109.2.
